### ISSUE TITLE
Logstash-1.5.0.beta1 and log-courier 1.3 sending repeated payloads, but no data coming through

### ISSUE THREAD

**Issue by jc21** [driskell/log-courier#103]
Hi,

I'm having trouble actually getting log-courier to send anything to my logstash instance, using TLS or TCP. The logstash server is not processing any incoming messages from the courier input. Here's the client log:

	2015/01/27 11:31:24.359514 Log Courier version 1.3 pipeline starting
	2015/01/27 11:31:24.359726 Loading registrar data from ./.log-courier
	2015/01/27 11:31:24.360137 Pipeline ready
	2015/01/27 11:31:24.361407 Attempting to connect to 10.0.0.75:4400 (logstash.lan)
	2015/01/27 11:31:24.361637 Resuming harvester on a previously harvested file: /var/log/dstat-monitor/json.log
	2015/01/27 11:31:24.361715 Resuming harvester on a previously harvested file: /var/log/db-monitor/json.log
	2015/01/27 11:31:24.361859 Started harvester at position 17902339 (requested 17902339): /var/log/dstat-monitor/json.log
	2015/01/27 11:31:24.362438 Started harvester at position 909020 (requested 909020): /var/log/db-monitor/json.log
	...
	2015/01/27 11:38:50.593709 ACKN message received for payload 4c2b46098b8fb17487bfc3b7b62cde97 sequence 0
	2015/01/27 11:38:50.593784 10 payloads still pending, resetting timeout
	2015/01/27 11:38:55.597804 ACKN message received for payload 4c2b46098b8fb17487bfc3b7b62cde97 sequence 0
	2015/01/27 11:38:55.597926 10 payloads still pending, resetting timeout
	2015/01/27 11:39:00.601820 ACKN message received for payload 4c2b46098b8fb17487bfc3b7b62cde97 sequence 0
	2015/01/27 11:39:00.601886 10 payloads still pending, resetting timeout
	2015/01/27 11:39:05.606478 ACKN message received for payload 4c2b46098b8fb17487bfc3b7b62cde97 sequence 0
	2015/01/27 11:39:05.606543 10 payloads still pending, resetting timeout
	2015/01/27 11:39:10.610719 ACKN message received for payload 4c2b46098b8fb17487bfc3b7b62cde97 sequence 0
	2015/01/27 11:39:10.610785 10 payloads still pending, resetting timeout

There are no log messages or errors on the server side. Looking at another closed issue it suggests that the logstash server is stalled, but this isn't true in my case. Other inputs are working. I think it's the courier input that might not be working as expected?

Logstash Server:

	input {
		courier {
			port      => 4400
			transport => "tcp"
		}
	}

	output {
		elasticsearch {
			host     => "127.0.0.1"
			port     => 9200
			protocol => "http"
		}
	}

Log-courier client:

	{
		"general": {
			"admin enabled": true,
			"log file": "/var/log/log-courier",
			"log level": "debug",
			"host": "www.lan"
		},
		"network": {
			"servers": [ "logstash.lan:4400" ],
			"transport": "tcp",
			"reconnect": 5
		},
		"files": [
			{
				"paths": [ "/var/log/dstat-monitor/json.log", "/var/log/db-monitor/json.log" ],
				"fields": { "type": "dstat" }
			}
		]
	}

Any ideas?

**Comment by vespian** [comment]
Bump.

We have hit the same issue here :(

**Comment by driskell** [comment]
Can you run Logstash in debug mode and monitor the output?

**Comment by vespian** [comment]
Thanks for a reply.

Yes, I can but the amount of data logged is quite a lot. What I should be looking for ?

**Comment by driskell** [comment]
@vespian The OP stated no logs on the server side and absolutely nothing is processed. Is this the same in your instance? Or is your issue something that develops over time. Best to create new issue if it is.

If it's the same, logging should be producing a lot since nothing is processing. Might be worth seeing what it is actually getting logged in this case. If the issue is a developing one though, this can still help - just need to run in debug until it happens, then look at where in the logs it stops, or if the logs still are showing something but just slower for example.

Generally though, these kind of issues are nearly always a problem on the output side. @jc21 's logs show the server is responding partial acknowledgements saying it's waiting on logstash pipeline so the log-courier to logstash communication is all working - just events aren't moving. Of course, we can't rule out completely that there is an issue on the input side - it's just unlikely

**Comment by vespian** [comment]
I have isolated logstash to just one log-courier client. I can see messages being processed after restarting logstash, and while log-courier outputs this:

```
Jan 28 23:05:21 front log-courier[28080]: ACKN message received for payload 3f3df18c26d6e049bf0538744b5562de sequence 44
Jan 28 23:05:46 front log-courier[28080]: 4 payloads still pending, resetting timeout
```

logstash is busy outputting this:

```
{:timestamp=>"2015-01-29T00:02:30.964000+0100", :message=>"Partially acknowledging message", :nonce=>"3f3df18c26d6e049bf0538744b5562de", :sequence=>44, :level=>:debug, :file=>"log-courier/server.rb", :line=>"193", "plugin"=>"input/courier", "port"=>6782}
```

I will continue debbuging it later on.

**Comment by jc21** [comment]
I've done a test using a fresh install of centos, logstash and log-courier, taking via tcp to localhost and outputting to a file. Test was all good. Will do some more testing with other situations.

**Comment by jc21** [comment]
- Local log-courier to logstash with file output:   OK
- Local log-courier to logstash with ES output:     OK
- Remote log-courier to logstash with file output:  OK
- Remote log-courier to logstash with ES output:    OK



I think I've found the problem. I have a mutate to float filter:

	filter {
		mutate {
			convert => [ "stat_value", "float" ]
		}
	}
		
which is causing this Logstash error on startup:

	Exception in filterworker {"exception"=>#<NoMethodError: undefined method `to_f' for #<Java::JavaMath::BigDecimal:0x86a8fa1>>, "backtrace"=>["/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-mutate-0.1.2/lib/logstash/filters/mutate.rb:290:in `convert_float'", "org/jruby/RubyMethod.java:128:in `call'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-mutate-0.1.2/lib/logstash/filters/mutate.rb:271:in `convert'", "org/jruby/RubyHash.java:1341:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-mutate-0.1.2/lib/logstash/filters/mutate.rb:256:in `convert'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-mutate-0.1.2/lib/logstash/filters/mutate.rb:210:in `filter'", "(eval):206:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):202:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "org/jruby/RubyEnumerable.java:805:in `flat_map'", "(eval):178:in `initialize'", "org/jruby/RubyProc.java:271:in `call'", "/opt/logstash/lib/logstash/pipeline.rb:268:in `filter'", "/opt/logstash/lib/logstash/pipeline.rb:206:in `filterworker'", "/opt/logstash/lib/logstash/pipeline.rb:150:in `start_filters'"], :level=>:error, "plugin"=>"input/courier", "port"=>4400}

So is that a courier issue or a logstash 1.5 issue?

**Comment by driskell** [comment]
@jc21 Could be either! I'll investigate. Courier plugin does fast JSON unfolding that creates "Rubylike" Java objects. I'll try reproduce and see what the json codec does. Thanks

**Comment by driskell** [comment]
@jc21 This might be an issue in Logstash. I'll investigate. I should be able to fix it in the courier plugins but be good to fix it in Logstash overall.

Can you tell me where stat_value comes from though? As in what your log courier config is and what the input config is and any filters between the input and broken filter? Just need to work out where the BigDecimal appears. Thanks

**Comment by jc21** [comment]
Sure. Here's the log-courier.conf:

	{
		"general": {
			"admin enabled": true,
			"log file": "/var/log/log-courier",
			"host": "www.lan"
		},
		"network": {
			"servers": [ "logstash.lan:4400" ],
			"transport": "tcp",
			"reconnect": 5
		},
		"files": [
			{
				"paths": [ "/var/log/dstat-monitor/json.log", "/var/log/db-monitor/json.log" ],
				"fields": { "type": "dstat" }
			}
		]
	}

And here's the logstash.conf:

	input {
		courier {
			port      => 4400
			transport => "tcp"
		}
	}

	filter {
		if [type] == "dstat" {
			json {
				source => "message"
			}

			# handle duplicate events by creating a sha256 checksum of the message
			anonymize {
				key          => "@version"
				fields       => [ "@uuid" ]
				algorithm    => "SHA256"
			}

			date {
				locale       => "en"
				match        => [ "timestamp", "UNIX" ]
				target       => "@timestamp"
				remove_field => [ "timestamp" ]
			}

			mutate {
				convert      => [ "stat_value", "float" ]
				remove_field => [ "message" ]
			}
		}
	}

	output {
		elasticsearch {
			host     => "127.0.0.1"
			port     => 9200
			protocol => "http"
			index    => "dstat-%{+YYYY.MM.dd}"
		}
	}

And here's some log lines to test with:

	{"timestamp":1422837694,"stat_name":"procs_new","stat_value":15}
	{"timestamp":1422837694,"stat_name":"mem_used","stat_value":12777527705.6}
	{"timestamp":1422837694,"stat_name":"mem_buffered","stat_value":4392484864}
	{"timestamp":1422837694,"stat_name":"mem_cached","stat_value":15247133900.8}
	{"timestamp":1422837694,"stat_name":"mem_free","stat_value":1294991360}
	{"timestamp":1422837694,"stat_name":"swap_used","stat_value":835715072}
	{"timestamp":1422837694,"stat_name":"swap_free","stat_value":3268608}
	{"timestamp":1422837694,"stat_name":"load_1m","stat_value":1.61}
	{"timestamp":1422837694,"stat_name":"load_5m","stat_value":2.34}
	{"timestamp":1422837694,"stat_name":"load_15m","stat_value":4.43}

I also have a very complex logstash config on another server that just won't work with the 1.5 beta at all, even before trying log-courier with it. There were no errors output by logstash in verbose mode, there's something going on internally that just isn't ready for consumption.

**Comment by driskell** [comment]
I have a fix ready.

You could test it by patching the plugin in your Logstash installation.
If you can look for server.rb for "log-courier" gem and test the patch that will be great.

**Comment by mareklobaza** [comment]
Back to problem reported by @jc21 and @vespian. Experiencing the same but got finnaly exception producing it, unfortunatelly say nothing for me, maybe you can help!

Symptoms:
Log-courier all the time producing the same "ACKN message received for payload" for the same message and sequence number. Happens times to times. Only logstash restart helps, but first log-courier has to be stopped - seems log-courier held some resorces (zeromq? is even possible?)

`"transport": "plainzmq"`
Nothing in log:
```
{:timestamp=>"2015-04-24T13:21:12.510000+0200", :message=>"filters/LogStash::Filters::Xml: adding tag", :tag=>"xml_parsed", :level=>:debug, :file=>"logstash/filters/base.rb", :line=>"187", :method=>"filter_matched", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:18.848000+0200", :message=>"Idle source shutting down", :source=>"006b8b4568", :level=>:info, :file=>"log-courier/server_zmq.rb", :line=>"314", :method=>"try_drop", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:35.757000+0200", :message=>"Idle source shutting down", :source=>"006b8b4567", :level=>:info, :file=>"log-courier/server_zmq.rb", :line=>"314", :method=>"try_drop", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:38.841000+0200", :message=>"New source", :source=>"006b8b4568", :level=>:info, :file=>"log-courier/server_zmq.rb", :line=>"257", :method=>"deliver", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:40.594000+0200", :message=>"New source", :source=>"006b8b4567", :level=>:info, :file=>"log-courier/server_zmq.rb", :line=>"257", :method=>"deliver", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:43.866000+0200", :message=>"Partially acknowledging message", :nonce=>"41472f2199c3d4e7652c66272ed81a9b", :sequence=>13, :level=>:debug, :file=>"log-courier/server.rb", :line=>"193", :method=>"process_jdat", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:48.867000+0200", :message=>"Partially acknowledging message", :nonce=>"1e1d4fa553f6ec3d904f7344d5474d74", :sequence=>0, :level=>:debug, :file=>"log-courier/server.rb", :line=>"193", :method=>"process_jdat", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:53.867000+0200", :message=>"Partially acknowledging message", :nonce=>"41472f2199c3d4e7652c66272ed81a9b", :sequence=>13, :level=>:debug, :file=>"log-courier/server.rb", :line=>"193", :method=>"process_jdat", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:58.869000+0200", :message=>"Partially acknowledging message", :nonce=>"1e1d4fa553f6ec3d904f7344d5474d74", :sequence=>0, :level=>:debug, :file=>"log-courier/server.rb", :line=>"193", :method=>"process_jdat", "plugin"=>"input/courier", "port"=>5509}
{:timestamp=>"2015-04-24T13:21:58.869000+0200", :message=>"Partially acknowledging message", :nonce=>"41472f2199c3d4e7652c66272ed81a9b", :sequence=>13, :level=>:debug, :file=>"log-courier/server.rb", :line=>"193", :method=>"process_jdat", "plugin"=>"input/courier", "port"=>5509}
```
but there's exceptions produced on console:
```
Apr 24, 2015 1:20:16 PM org.elasticsearch.discovery.DiscoveryService doStart
INFO: [logstash-servername-17029-14950] DefaultCluster/mqC681fbSUCGe68J2QfzhQ
Apr 24, 2015 1:20:19 PM org.elasticsearch.cluster.service.InternalClusterService$UpdateTask run
INFO: [logstash-servername-17029-14950] detected_master (...)
Apr 24, 2015 1:20:19 PM org.elasticsearch.node.internal.InternalNode start
INFO: [logstash-servername-17029-14950] started
Exception in thread "|worker" java.lang.UnsupportedOperationException
        at java.lang.Thread.stop(Thread.java:869)
        at org.jruby.RubyThread.exceptionRaised(RubyThread.java:1221)
        at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:112)
        at java.lang.Thread.run(Thread.java:745)
```

If I restart logstash without stopping log-courier exception will be thrown imediatelly:
```
INFO: [logstash-servername-17029-14950] DefaultCluster/T9XNM033RH6xAcZf8yvt2A
Exception in thread "|worker" java.lang.UnsupportedOperationException
        at java.lang.Thread.stop(Thread.java:869)
        at org.jruby.RubyThread.exceptionRaised(RubyThread.java:1221)
        at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:112)
        at java.lang.Thread.run(Thread.java:745)
Apr 24, 2015 1:18:23 PM org.elasticsearch.cluster.service.InternalClusterService$UpdateTask run
INFO: [logstash-servername-17029-14950] detected_master (...)
Apr 24, 2015 1:18:23 PM org.elasticsearch.node.internal.InternalNode start
INFO: [logstash-servername-17029-14950] started
```
gives misleading view that everything is fine! Logstash will run with high CPU consumption doing nothing (only logging `Partially acknowledging message` in loop)!
Any idea?

**Comment by driskell** [comment]
@mareklobaza I believe your issue might be new and unrelated to the specific issue in this thread.

Some questions if you can raise a new issue with answers:
Version of Logstash and Log Courier and the courier gems
Are you balancing across multiple Logstash servers? How many?
Does this occur on them all?
Does it occur at random?
Can you provide configs?

The partially acknowledging message means the pipeline has stopped and the courier plugin is just keeping the connection alive by resending its last ACK. This is normal but usually only momentarily. The pipeline blocking completely isn't normal though - not sure why your filter worker is crashing too but its not for the reason in this thread.

**Comment by mareklobaza** [comment]
OK, created #155 with more details. 
You're right that partially acknowledging message is fine but only if  sequence number is increased. Logstash has nothing to do but won't processing message (sequence is the same) because worker thread is dead silently!
The same I can see in the attached logs, that's why I posted here.
By the way - thanks for really fast answer - appreciate ;)
